src/foam/nanos/dig/HTTPDigestSink.java

```diff
@@ -73,9 +75,57 @@ public HTTPDigestSink(String url, String bearerToken, DUGDigestConfig dugDigestC
 
   @Override
   public void put(Object obj, Detachable sub) {
+    try {
+      int responseCode = sendRequest(obj);
+
+      if ( responseCode == HttpServletResponse.SC_OK ) return;
+
+      if ( responseCode == HttpServletResponse.SC_BAD_REQUEST ) { // client error
+        if ( obj instanceof EmailWebhook ) {
+          EmailWebhook ewh = (EmailWebhook) obj ;
+          DAO alarmDAO = (DAO) getX().get("alarmDAO");
+          String emailAddress = ewh.getEmail();
+          String note = "[" + emailAddress + ", " + new Date() + "]";
+          String name = "EmailWebhook";
+          Alarm alarm = (Alarm) alarmDAO.find(EQ(Alarm.NAME, name));
+
+          if ( alarm == null ) {
+            alarm = new Alarm.Builder(getX())
+              .setName(name)
+              .setSeverity(LogLevel.ERROR)
+              .setNote("Email not sent to:\n\t" + note)
+              .build();
+          } else {
+            alarm.setNote(alarm.getNote() + "\n\t" + note);
+          }
+          alarmDAO.put(alarm);
```

Comment:
we should log too

Reply:
We throw an exception everywhere we create an alarm and that logs the result for us.

![Uploading Screen Shot 2022-10-05 at 11.33.56 AM.png…]()